compiler/rustc_codegen_llvm/src/mono_item.rs

```diff
@@ -110,13 +114,25 @@ impl CodegenCx<'ll, 'tcx> {
             return true;
         }
 
+        // PowerPC64 prefers TOC indirection to avoid copy relocations.
+        let arch = &self.tcx.sess.target.arch;
+        if arch == "powerpc64" || arch == "powerpc64le" {
+            return false;
+        }
+
+        // Thread-local variables generally don't support copy relocations.
+        if attrs.flags.contains(CodegenFnAttrFlags::THREAD_LOCAL) {
+            return false;
+        }
+
         // Static relocation model should force copy relocations everywhere.
         if self.tcx.sess.relocation_model() == RelocModel::Static {
             return true;
         }
 
         // Symbols from executables can't really be imported any further.
-        if self.tcx.sess.crate_types().iter().all(|ty| *ty == CrateType::Executable) {
+        let all_exe = self.tcx.sess.crate_types().iter().all(|ty| *ty == CrateType::Executable);
+        if !is_declaration && all_exe {
```

Comment:
As per clang's code what matters is not whether this is a declaration but whether it is a declaration for _linker_. There's one additional case where this is true – when the definition has an AvailableExternally linkage.

Reply:
Additionally, from Clang's code it seems that this kind of check is done before the ppc64 one. IMO we should try to mirror it.

Reply:
Addressed both points in the new version.